src/Steps/Release/MergeUpRelease.php

```diff
@@ -0,0 +1,181 @@
+            '/' . preg_quote(RewriteReleaseBranches::RELEASE_SUFFIX) . '$/',
+            $currentBranch
+        );
+        if (!$currentBranchIsRelease) {
+            $this->log(
+                $output,
+                "Warning: Current branch '$currentBranch' is not a release branch. "
+                    . "Skipping $currentLibrary and its children.",
+                'bg=yellow'
+            );
+            return;
+        }
+
+        // Do the merge up recursively
+        foreach ($releasePlanNode->getItems() as $item) {
+            $this->mergeRecursive($output, $item, $deleteBranch);
+        }
+        $this->mergeLibrary($output, $releasePlanNode);
+
+        // Optionally delete the branch from the remote
+        if ($deleteBranch) {
+            $this->log(
+                $output,
+                "Deleting branch <info>$currentBranch</info> from remote for <info>{$currentLibrary}</info>"
+            );
+            $releasePlanNode->getLibrary()->getRepository()->run('push', ['--delete', $currentBranch]);
+        }
+    }
+
+    /**
+     * Performs a merge up for a single library
+     *
+     * @param OutputInterface $output
+     * @param LibraryRelease $releasePlanNode Node in release plan being released
+     */
+    protected function mergeLibrary(OutputInterface $output, LibraryRelease $releasePlanNode)
+    {
+        $library = $releasePlanNode->getLibrary();
+        $releaseBranch = $library->getBranch();
+        $minorBranch = str_replace(RewriteReleaseBranches::RELEASE_SUFFIX, '', $releaseBranch);
+        $majorBranch = $releasePlanNode->getVersion()->getMajor();
+
+        // Merge up to minor branch, then to major branch
+        $this->mergeUp($output, $releasePlanNode, $releaseBranch, $minorBranch, false);
+        $this->mergeUp($output, $releasePlanNode, $minorBranch, $majorBranch, true);
+
+        // Always check out the release branch at the end
+        $library->checkout($output, $releaseBranch);
+    }
+
+    /**
+     * Merge up from one branch to another, setting the appropriate constraints for released dependencies as we go.
+     */
+    protected function mergeUp(
+        OutputInterface $output,
+        LibraryRelease $releasePlanNode,
+        string $mergeFrom,
+        string $mergeInto,
+        string $isMajorBranch
+    ) {
+        $library = $releasePlanNode->getLibrary();
+        $name = $library->getName();
+
+        // Step 1. Checkout branch to merge into
+        $library->checkout($output, $mergeInto);
+
+        $this->log(
+            $output,
+            "Merging library <info>{$name}</info> from <info>{$mergeFrom}</info> to <info>{$mergeInto}</info>"
+        );
+
+        // Step 2. Merge up if there's anything to merge
+        if ($library->hasDiff($output, $mergeFrom)) {
+            // Note: If there are conflicts or other issues merging,
+            // a Gitonomy\Git\Exception\ProcessException will be thrown and cow will exit
```

Comment:
I would be shocked if someone thought adding a global exception handler to cow was a good idea, it uses exceptions to bail in many places, so that would already cause problems.

The exception already includes the git error message, which in the case of a commit will look like this (obviously a conflict in composer.json is unlikely but this gets the idea across):
> Auto-merging composer.json
> CONFLICT (content): Merge conflict in composer.json
> Automatic merge failed; fix conflicts and then commit the result.

I don't think we need to be any more explicit about what commands to use than this - anyone doing a release already knows how to resolve merge conflicts. Besides which it's concievable that the merge could fail for other reasons other than a merge conflict (though I admittedly can't think of such a scenario right now) - in which case the exception will say what went wrong, but adding messaging saying to resolve a merge conflict that doesn't exist would be confusing.

I'll add more explicit messaging if you want, and I'll catch the exception and call `die()` instead if you want, but I don't think either of those are necessary.

Reply:
I think we still need the explict messaging with instructions.  Cow is a bit of black box and the person managing the release may not fully grasp the 3x branches and stable vs dev dependencies in recipes (I know it seems obvious now, but it took me multiple releases to really get my head around this, though to be fair most of the confusion was because of --detach-tagged-base)

For instance, simply saying `git merge --no-ff 1.12-release` is going to be valuable in this context so the user knows they're merge from the correct branch.

Reply:
Cow will already log out "Merging library <info>{$name}</info> from <info>{$mergeFrom}</info> to <info>{$mergeInto}</info>"
This will result in something like:

> Merging library `silverstripe/framework` from `4.12-release` to `4.12`

Then the exception message from git will say (along with some debug info) the message in my previous comment above. So the context for what repo is being dealt with, what went wrong, what branch is being merged into what other branch, that's all there already.
When checking out the directory for that repo, you'll be in the conflict state already so running `git merge --no-ff 1.12-release` is unnecessary.

I'm adding messaging for what commands to run after resolving the conflict as we discussed. It's concievable that there will be some scenario where this is _not_ the correct set of commands to run, but since I can't think of a specific scenario I'll just let it be.